File: ASTStatisticsGenerator.js
// To run this script:
// 1. First, generate the ASTs using the previous script.
// 2. Make sure the ASTs are in a directory (e.g., 'JavascriptAST').
// 3. Execute this script from your terminal, passing the path to the AST directory:
//    node analyze.js ./JavascriptAST
// 4. A file named 'analysis_report.json' will be created in the current directory.

const fs = require('fs');
const path = require('path');

// --- Helper Functions for Analysis ---

/**
 * Traverses the AST to find all nodes of a specific type.
 * @param {object} node - The current AST node to start from.
 * @param {string} type - The node type to find (e.g., 'function_declaration').
 * @returns {object[]} An array of nodes that match the type.
 */
function findNodesByType(node, type) {
    let nodes = [];
    if (!node) return nodes;

    if (node.type === type) {
        nodes.push(node);
    }

    if (node.children && node.children.length > 0) {
        for (const child of node.children) {
            nodes = nodes.concat(findNodesByType(child, type));
        }
    }
    return nodes;
}

/**
 * Calculates the cyclomatic complexity of a function node.
 * @param {object} functionNode - The AST node of a function.
 * @returns {number} The cyclomatic complexity score.
 */
function calculateCyclomaticComplexity(functionNode) {
    let complexity = 1;
    const branchingTypes = [
        'if_statement', 'for_statement', 'while_statement', 'case_statement',
        'catch_clause', 'ternary_expression'
    ];
    
    findNodesByType(functionNode, 'binary_expression').forEach(node => {
        if (node.children.some(c => c.type === '&&' || c.type === '||')) {
            complexity++;
        }
    });

    branchingTypes.forEach(type => {
        complexity += findNodesByType(functionNode, type).length;
    });

    return complexity;
}

/**
 * Analyzes a single AST file and aggregates statistics.
 * @param {string} filePath - Path to the AST JSON file.
 * @param {object} stats - The global statistics object to update.
 */
function analyzeAstFile(filePath, stats) {
    const jsonContent = fs.readFileSync(filePath, 'utf8');
    const ast = JSON.parse(jsonContent);
    if (!ast) return;

    // --- 1. Code Composition & Complexity ---
    stats.composition.fileCount++;
    const linesOfCode = ast.endPosition.row - ast.startPosition.row + 1;
    stats.composition.totalLinesOfCode += linesOfCode;
    stats.composition.linesOfCodePerFile[filePath] = linesOfCode;

    const functions = findNodesByType(ast, 'function_declaration').concat(findNodesByType(ast, 'arrow_function'), findNodesByType(ast, 'method_definition'));
    stats.composition.functionCount += functions.length;
    stats.composition.classCount += findNodesByType(ast, 'class_declaration').length;
    
    const comments = findNodesByType(ast, 'comment');
    stats.composition.totalComments += comments.length;
    
    let maxDepth = 0;
    function getDepth(node, depth) {
        if (!node) return;
        maxDepth = Math.max(maxDepth, depth);
        if(node.children) node.children.forEach(child => getDepth(child, depth + 1));
    }
    getDepth(ast, 0);
    stats.complexity.maxNestingDepth = Math.max(stats.complexity.maxNestingDepth, maxDepth);

    let totalComplexity = 0;
    for (const func of functions) {
        totalComplexity += calculateCyclomaticComplexity(func);
    }
    stats.complexity.totalCyclomatic += totalComplexity;
    
    // --- 2. Dependency Analysis ---
    const imports = findNodesByType(ast, 'import_statement');
    stats.dependencies.importCount += imports.length;
    for (const imp of imports) {
        const sourceNode = imp.children.find(c => c.type === 'string');
        if (sourceNode) {
            const source = sourceNode.text.replace(/['"]/g, '');
            stats.dependencies.importFrequency[source] = (stats.dependencies.importFrequency[source] || 0) + 1;
            if (source.startsWith('.') || source.startsWith('@/')) {
                stats.dependencies.internalImports++;
            } else {
                stats.dependencies.externalImports++;
            }
             // Check for cloud SDKs within dependency analysis
            if (source.includes('@aws-sdk')) stats.infra.cloudSDKs.add('AWS');
            if (source.includes('@google-cloud')) stats.infra.cloudSDKs.add('Google Cloud');
            if (source.includes('firebase')) stats.infra.cloudSDKs.add('Firebase');
        }
    }

    // --- 3. API and Service Usage ---
    const apiCallKeywords = ['fetch', 'axios', 'http', 'https'];
    const callExpressions = findNodesByType(ast, 'call_expression');
    for (const call of callExpressions) {
         if (apiCallKeywords.some(kw => call.text.toLowerCase().includes(kw))) {
            stats.api.networkCallCount++;
        }
         // React Hook Detection
        const identifier = call.children[0]?.text;
        if (identifier && stats.reactSpecifics.hooks[identifier] !== undefined) {
            stats.reactSpecifics.hooks[identifier]++;
        }
    }
    const hardcodedUrlRegex = /https?:\/\/[^\s/$.?#].[^\s]*/gi;
    findNodesByType(ast, 'string').forEach(node => {
        const matches = node.text.match(hardcodedUrlRegex);
        if(matches){
            stats.api.hardcodedUrls.push(...matches);
        }
    });
    stats.api.fileIOCount += findNodesByType(ast, 'call_expression').filter(c => c.text.startsWith('fs.')).length;
    
    // --- 4. Code Quality & Maintainability ---
    stats.quality.tryCatchCount += findNodesByType(ast, 'try_statement').length;
    for (const comment of comments) {
        if (comment.text.toLowerCase().includes('todo') || comment.text.toLowerCase().includes('fixme')) {
            stats.quality.todoFixmeCount++;
        }
    }
    if (filePath.includes('.test.') || filePath.includes('.spec.')) {
        stats.quality.testFileCount++;
    }

    // --- 5. Application Structure ---
    const dir = path.dirname(filePath).split(path.sep).pop() || 'root';
    stats.structure.moduleSize[dir] = (stats.structure.moduleSize[dir] || 0) + linesOfCode;
    
    // --- 6. Modern JS & Code Style ---
    findNodesByType(ast, 'lexical_declaration').forEach(node => {
        if (node.text.startsWith('const')) stats.modernFeatures.variableDeclarations.const++;
        if (node.text.startsWith('let')) stats.modernFeatures.variableDeclarations.let++;
    });
    stats.modernFeatures.variableDeclarations.var += findNodesByType(ast, 'variable_declaration').length;
    stats.modernFeatures.arrowFunctions += findNodesByType(ast, 'arrow_function').length;
    stats.modernFeatures.templateLiterals += findNodesByType(ast, 'template_string').length;
    stats.modernFeatures.destructuringAssignments += findNodesByType(ast, 'object_pattern').length + findNodesByType(ast, 'array_pattern').length;
    stats.modernFeatures.spreadOperators += findNodesByType(ast, 'spread_element').length;
    
    // --- 7. Framework Specifics (React & Angular) ---
    const classDeclarations = findNodesByType(ast, 'class_declaration');
    for (const classNode of classDeclarations) {
        if (classNode.children.some(c => c.type === 'class_heritage' && c.text.includes('Component'))) {
            stats.reactSpecifics.classComponents++;
        }
        if (classNode.children.some(c => c.type === 'decorator' && c.text.startsWith('@Component'))) {
            stats.angularSpecifics.components++;
        }
    }
    if(findNodesByType(ast, 'jsx_element').length > 0 && !classDeclarations.some(c => c.children.some(child => child.type === 'class_heritage' && child.text.includes('Component')))){
         stats.reactSpecifics.functionalComponents++;
    }

    // --- 8. Infrastructure & Security ---
    findNodesByType(ast, 'member_expression').forEach(node => {
        if (node.text.startsWith('process.env')) {
            stats.infra.environmentVariables.add(node.text);
        }
    });
    findNodesByType(ast, 'jsx_attribute').forEach(attr => {
        if (attr.children[0]?.text === 'dangerouslySetInnerHTML') {
            stats.security.dangerouslySetInnerHTML++;
        }
        if (attr.children.some(c => c.type === 'jsx_expression' && findNodesByType(c, 'arrow_function').length > 0)) {
            stats.security.anonymousFuncInProps++;
        }
    });
}

// --- Main Execution ---
function main(astDir) {
    if (!fs.existsSync(astDir)) {
        console.error(`Error: Directory not found at '${astDir}'`);
        return;
    }

    const stats = {
        composition: { fileCount: 0, functionCount: 0, classCount: 0, totalLinesOfCode: 0, totalComments: 0, linesOfCodePerFile: {} },
        complexity: { maxNestingDepth: 0, totalCyclomatic: 0 },
        dependencies: { importCount: 0, internalImports: 0, externalImports: 0, importFrequency: {} },
        api: { networkCallCount: 0, fileIOCount: 0, hardcodedUrls: [] },
        quality: { todoFixmeCount: 0, tryCatchCount: 0, testFileCount: 0 },
        structure: { moduleSize: {} },
        modernFeatures: {
            variableDeclarations: { const: 0, let: 0, var: 0 },
            arrowFunctions: 0,
            templateLiterals: 0,
            destructuringAssignments: 0,
            spreadOperators: 0,
        },
        reactSpecifics: {
            classComponents: 0,
            functionalComponents: 0,
            hooks: { useState: 0, useEffect: 0, useContext: 0, useReducer: 0, useCallback: 0, useMemo: 0, useRef: 0 },
        },
        angularSpecifics: {
            components: 0,
        },
        infra: {
            environmentVariables: new Set(),
            cloudSDKs: new Set(),
        },
        security: {
            dangerouslySetInnerHTML: 0,
            anonymousFuncInProps: 0,
        }
    };

    function processDirectory(directory) {
        const files = fs.readdirSync(directory, { withFileTypes: true });
        for (const file of files) {
            const fullPath = path.join(directory, file.name);
            if (file.isDirectory()) {
                processDirectory(fullPath);
            } else if (path.extname(file.name) === '.json') {
                analyzeAstFile(fullPath, stats);
            }
        }
    }

    processDirectory(astDir);
    saveReportAsJson(stats);
}

/**
 * Saves the final statistics report to a JSON file.
 * @param {object} stats - The fully aggregated statistics object.
 */
function saveReportAsJson(stats) {
    // Convert sets to arrays for JSON serialization
    stats.infra.environmentVariables = Array.from(stats.infra.environmentVariables);
    stats.infra.cloudSDKs = Array.from(stats.infra.cloudSDKs);
    const externalDeps = Object.keys(stats.dependencies.importFrequency).filter(k => !k.startsWith('.') && !k.startsWith('@/'));

    const report = {
        "Code Composition & Complexity": {
            "Number of files": stats.composition.fileCount,
            "Total lines of code": stats.composition.totalLinesOfCode,
            "Number of functions": stats.composition.functionCount,
            "Number of classes": stats.composition.classCount,
            "Maximum nesting depth": stats.complexity.maxNestingDepth,
            "Total cyclomatic complexity": stats.complexity.totalCyclomatic,
            "Average cyclomatic complexity": parseFloat((stats.complexity.totalCyclomatic / (stats.composition.functionCount || 1)).toFixed(2)),
            "Total number of comments": stats.composition.totalComments,
            "Comment-to-code ratio": parseFloat((stats.composition.totalComments / (stats.composition.totalLinesOfCode || 1)).toFixed(3))
        },
        "Dependency Analysis": {
            "Number of import statements": stats.dependencies.importCount,
            "Types of imports (internal vs. external)": `${stats.dependencies.internalImports} vs. ${stats.dependencies.externalImports}`,
            "Third-party vs. in-house code ratio": parseFloat((stats.dependencies.externalImports / (stats.dependencies.importCount || 1)).toFixed(2)),
            "List of direct dependencies": externalDeps,
            "Import frequency by module/library": stats.dependencies.importFrequency
        },
        "API and Service Usage": {
            "Total number of network/API calls (heuristic)": stats.api.networkCallCount,
            "Total number of file I/O operations": stats.api.fileIOCount,
            "List and count of hardcoded URLs": {
                count: stats.api.hardcodedUrls.length,
                urls: stats.api.hardcodedUrls
            }
        },
        "Code Quality & Maintainability": {
            "Number of TODO / FIXME comments": stats.quality.todoFixmeCount,
            "Number of exception handling blocks (try/catch)": stats.quality.tryCatchCount,
            "Number of test files": stats.quality.testFileCount
        },
        "Application Structure": {
            "Number of modules/packages": Object.keys(stats.structure.moduleSize).length,
            "Size of each module/package (in lines of code)": stats.structure.moduleSize
        },
        "Modern Javascript Usage": stats.modernFeatures,
        "Framework Specifics": {
            "React": stats.reactSpecifics,
            "Angular": stats.angularSpecifics
        },
        "Infrastructure & Security": {
            "Environment Variables Accessed": stats.infra.environmentVariables,
            "Cloud SDKs Detected": stats.infra.cloudSDKs,
            "Use of dangerouslySetInnerHTML": stats.security.dangerouslySetInnerHTML,
            "Anonymous Functions in Props": stats.security.anonymousFuncInProps
        },
        "Unsupported or Partially Supported Metrics": {
            "Test coverage": "N/A from AST alone; needs runtime/test suite execution"
        }
    };

    const outputFilePath = path.join(process.cwd(), 'analysis_report.json');
    try {
        fs.writeFileSync(outputFilePath, JSON.stringify(report, null, 2));
        console.log(`\n✅ Successfully saved analysis report to: ${outputFilePath}`);
    } catch (e) {
        console.error(`\n❌ Failed to write report to file. Reason: ${e.message}`);
    }
}


// --- Entry Point ---
const astDirectory = process.argv[2];
if (!astDirectory) {
    console.error("Please provide the path to the directory containing the AST JSON files.");
    console.log("Usage: node analyze.js <path-to-ast-directory>");
} else {
    main(astDirectory);
}

File: ASTStatisticsGenerator.py
# To run this script:
# 1. First, generate the ASTs using the python_ast_generator.py script.
# 2. Make sure the ASTs are in a directory named 'PythonAST'.
# 3. Execute this script from your terminal, passing the path to the AST directory:
#    python analyze_python_ast.py ./PythonAST

import json
import os
import sys

# --- Helper Functions for Analysis ---

def find_nodes_by_type(node, type_name):
    """
    Recursively finds all nodes of a specific type in the AST.
    """
    nodes = []
    if not isinstance(node, dict) or not node:
        return nodes

    if node.get('type') == type_name:
        nodes.append(node)

    for child in node.get('children', []):
        nodes.extend(find_nodes_by_type(child, type_name))
        
    return nodes

def calculate_cyclomatic_complexity(function_node):
    """
    Calculates the cyclomatic complexity for a Python function node.
    """
    complexity = 1
    # Python-specific branching nodes
    branching_types = [
        'if_statement', 'for_statement', 'while_statement', 
        'except_clause', 'assert_statement', 'with_statement'
    ]
    # 'and' and 'or' in boolean operators also add to complexity
    logical_operators = find_nodes_by_type(function_node, 'boolean_operator')
    complexity += len(logical_operators)
    
    for branch_type in branching_types:
        complexity += len(find_nodes_by_type(function_node, branch_type))
        
    return complexity

def analyze_ast_file(file_path, stats):
    """
    Analyzes a single AST file and aggregates statistics based on observed structures.
    """
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            ast_data = json.load(f)
    except (json.JSONDecodeError, FileNotFoundError) as e:
        print(f"[ERROR] Could not read or parse {file_path}: {e}")
        return

    if not ast_data:
        return

    # --- 1. Code Composition & Complexity ---
    stats['composition']['fileCount'] += 1
    lines = ast_data.get('endPosition', {}).get('row', 0) - ast_data.get('startPosition', {}).get('row', 0) + 1
    stats['composition']['totalLinesOfCode'] += lines
    
    functions = find_nodes_by_type(ast_data, 'function_definition')
    stats['composition']['functionCount'] += len(functions)
    stats['composition']['classCount'] += len(find_nodes_by_type(ast_data, 'class_definition'))
    
    comments = find_nodes_by_type(ast_data, 'comment')
    stats['composition']['totalComments'] += len(comments)

    for func in functions:
        stats['complexity']['totalCyclomatic'] += calculate_cyclomatic_complexity(func)

    # --- 2. Dependency Analysis ---
    imports = find_nodes_by_type(ast_data, 'import_statement')
    imports_from = find_nodes_by_type(ast_data, 'import_from_statement')
    stats['dependencies']['importCount'] += len(imports) + len(imports_from)

    all_imports = imports + imports_from
    for imp in all_imports:
        source_node = next((c for c in imp.get('children', []) if c.get('type') in ['dotted_name', 'relative_import']), None)
        if source_node:
            source = source_node.get('text', '')
            stats['dependencies']['importFrequency'][source] = stats['dependencies']['importFrequency'].get(source, 0) + 1
            if source.startswith('.'):
                 stats['dependencies']['internalImports'] += 1
            else:
                 stats['dependencies']['externalImports'] += 1
                 if 'boto3' in source or 'google.cloud' in source:
                     stats['infra']['cloudSDKs'].add('AWS' if 'boto3' in source else 'Google Cloud')
                 if 'pymongo' in source:
                      stats['infra']['databaseTech'].add('MongoDB')
                 if 'flask' in source:
                      stats['frameworks']['detected'].add('Flask')
                 if 'django' in source:
                      stats['frameworks']['detected'].add('Django')


    # --- 3. API, Service, and Infrastructure Usage ---
    call_expressions = find_nodes_by_type(ast_data, 'call')
    for call in call_expressions:
        call_text = call.get('text', '')
        if any(keyword in call_text for keyword in ['requests.', 'httpx.', 'urllib.request']):
            stats['api']['networkCallCount'] += 1
        # Structural check for DB queries (more robust)
        if call.get('children'):
            function_call_node = call['children'][0]
            if function_call_node.get('type') == 'attribute' and any(db_call in function_call_node.get('text', '') for db_call in ['.query', '.execute', '.fetchone', '.fetchall', '.insert_one', '.find_one', '.add', '.commit']):
                 stats['api']['databaseQueries'] += 1
        
        if 'open(' in call_text:
            stats['api']['fileIOCount'] += 1
        if 'os.getenv' in call_text or 'os.environ' in call_text:
            arg_list = find_nodes_by_type(call, 'argument_list')
            if arg_list and arg_list[0].get('children'):
                 env_var_node = find_nodes_by_type(arg_list[0], 'string')
                 if env_var_node:
                    stats['infra']['environmentVariables'].add(env_var_node[0].get('text', ''))
    
    # --- 4. Code Quality & Maintainability ---
    stats['quality']['tryExceptCount'] += len(find_nodes_by_type(ast_data, 'try_statement'))
    for comment in comments:
        if 'TODO' in comment.get('text', '').upper() or 'FIXME' in comment.get('text', '').upper():
            stats['quality']['todoFixmeCount'] += 1
    if file_path.startswith('test') or file_path.endswith('_test.py') or 'test' in file_path:
        stats['quality']['testFileCount'] += 1
        
    # --- 5. Python-Specific & Framework Patterns ---
    stats['pythonSpecifics']['fStrings'] += len([s for s in find_nodes_by_type(ast_data, 'string') if s.get('text', '').startswith('f')])
    stats['pythonSpecifics']['listComprehensions'] += len(find_nodes_by_type(ast_data, 'list_comprehension'))
    
    # Structural check for Flask/Django routes
    decorated_defs = find_nodes_by_type(ast_data, 'decorated_definition')
    for dec_def in decorated_defs:
        for decorator_node in dec_def.get('children', []):
            if decorator_node.get('type') == 'decorator':
                dec_text = decorator_node.get('text', '')
                if '@app.route' in dec_text or '.route(' in dec_text:
                    stats['frameworks']['endpointsDefined'] += 1
                    # Extract endpoint path
                    route_call = find_nodes_by_type(decorator_node, 'call')
                    if route_call:
                        arg_list = find_nodes_by_type(route_call[0], 'argument_list')
                        if arg_list:
                            string_node = find_nodes_by_type(arg_list[0], 'string')
                            if string_node:
                                stats['api']['endpointPaths'].add(string_node[0].get('text'))
    
    # Heuristic for Django Models
    for class_node in find_nodes_by_type(ast_data, 'class_definition'):
        arg_list = find_nodes_by_type(class_node, 'argument_list')
        if arg_list and 'models.Model' in arg_list[0].get('text', ''):
             stats['frameworks']['djangoModels'] += 1


# --- Main Execution ---
def main(ast_dir):
    if not os.path.isdir(ast_dir):
        print(f"Error: Directory not found at '{ast_dir}'")
        sys.exit(1)

    # Initialize stats structure
    stats = {
        'composition': {'fileCount': 0, 'functionCount': 0, 'classCount': 0, 'totalLinesOfCode': 0, 'totalComments': 0},
        'complexity': {'totalCyclomatic': 0},
        'dependencies': {'importCount': 0, 'internalImports': 0, 'externalImports': 0, 'importFrequency': {}},
        'api': {'endpointsDefined': 0, 'endpointPaths': set(), 'networkCallCount': 0, 'databaseQueries': 0, 'fileIOCount': 0},
        'quality': {'todoFixmeCount': 0, 'tryExceptCount': 0, 'testFileCount': 0},
        'pythonSpecifics': {'fStrings': 0, 'listComprehensions': 0},
        'frameworks': {'detected': set(), 'endpointsDefined': 0, 'djangoModels': 0},
        'infra': {'environmentVariables': set(), 'databaseTech': set(), 'cloudSDKs': set()}
    }

    for root, _, files in os.walk(ast_dir):
        for file in files:
            if file.endswith('.json'):
                analyze_ast_file(os.path.join(root, file), stats)
    
    # --- Final Calculations & Formatting ---
    if stats['composition']['functionCount'] > 0:
        stats['complexity']['averageCyclomatic'] = round(stats['complexity']['totalCyclomatic'] / stats['composition']['functionCount'], 2)
    else:
        stats['complexity']['averageCyclomatic'] = 0
        
    stats['infra']['environmentVariables'] = list(stats['infra']['environmentVariables'])
    stats['infra']['cloudSDKs'] = list(stats['infra']['cloudSDKs'])
    stats['infra']['databaseTech'] = list(stats['infra']['databaseTech'])
    stats['frameworks']['detected'] = list(stats['frameworks']['detected'])
    stats['api']['endpointPaths'] = list(stats['api']['endpointPaths'])
    
    external_deps = {k: v for k, v in stats['dependencies']['importFrequency'].items() if not k.startswith('.')}
    
    # Structure the final report
    final_report = {
        "Code Composition & Complexity": {
            "Number of files": stats['composition']['fileCount'],
            "Total lines of code": stats['composition']['totalLinesOfCode'],
            "Number of functions": stats['composition']['functionCount'],
            "Number of classes": stats['composition']['classCount'],
            "Average cyclomatic complexity": stats['complexity']['averageCyclomatic'],
            "Total number of comments": stats['composition']['totalComments'],
        },
        "Dependency Analysis": {
            "Number of import statements": stats['dependencies']['importCount'],
            "Types of imports (internal vs. external)": f"{stats['dependencies']['internalImports']} vs. {stats['dependencies']['externalImports']}",
            "List of direct dependencies": list(external_deps.keys()),
            "Import frequency by module/library": stats['dependencies']['importFrequency'],
        },
        "API and Service Usage": {
            "Endpoints Defined": stats['frameworks']['endpointsDefined'],
            "Endpoint Paths": stats['api']['endpointPaths'],
            "Database Queries (heuristic)": stats['api']['databaseQueries'],
            "Outbound Network Calls (heuristic)": stats['api']['networkCallCount'],
            "File I/O Operations": stats['api']['fileIOCount'],
        },
        "Code Quality & Maintainability": {
            "Number of TODO / FIXME comments": stats['quality']['todoFixmeCount'],
            "Number of try/except blocks": stats['quality']['tryExceptCount'],
            "Number of test files": stats['quality']['testFileCount'],
        },
        "Python-Specific Patterns": stats['pythonSpecifics'],
        "Framework & Infrastructure": {
            "Frameworks Detected": stats['frameworks']['detected'],
            "Django Models Defined": stats['frameworks']['djangoModels'],
            "Environment Variables Accessed": stats['infra']['environmentVariables'],
            "Database Technologies Detected": stats['infra']['databaseTech'],
            "Cloud SDKs Detected": stats['infra']['cloudSDKs'],
        }
    }

    # Save report
    output_path = os.path.join(os.getcwd(), 'python_analysis_report.json')
    with open(output_path, 'w', encoding='utf-8') as f:
        json.dump(final_report, f, indent=2)
        
    print(f"\n✅ Successfully saved Python analysis report to: {output_path}")

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python analyze_python_ast.py <path-to-PythonAST-directory>")
        sys.exit(1)
    
    ast_directory = sys.argv[1]
    main(ast_directory)
